program/src/processor/initialize_config.rs

```diff
@@ -1,21 +1,137 @@
 ///   3. `[]` mint
-///   4. `[]` vault_token
+///   4. `[]` vault
 pub fn process_initialize_config(
-    _program_id: &Pubkey,
-    _ctx: Context<InitializeConfigAccounts>,
-    _cooldown_time_seconds: u64,
-    _max_deactivation_basis_points: u16,
+    program_id: &Pubkey,
+    ctx: Context<InitializeConfigAccounts>,
+    cooldown_time_seconds: u64,
+    max_deactivation_basis_points: u16,
 ) -> ProgramResult {
+    // Accounts validation.
+
+    // 1. mint
+    // - owner must the spl token 2022
+    // - must be initialized
+    // - have rewards transfer hook
+
+    require!(
+        ctx.accounts.mint.owner == &spl_token_2022::ID,
+        ProgramError::InvalidAccountOwner,
+        "mint"
+    );
+
+    let mint_data = ctx.accounts.mint.try_borrow_data()?;
+    // unpack checks if the mint is initialized
+    let mint = PodStateWithExtensions::<PodMint>::unpack(&mint_data)?;
+
+    // ensure the mint is configured with the expected `TransferHook` extension
+    if let Ok(transfer_hook) = mint.get_extension::<TransferHook>() {
+        let hook_program_id: Option<Pubkey> = transfer_hook.program_id.into();
+
+        require!(
+            hook_program_id == Some(REWARDS_PROGRAM_ID),
+            StakeError::InvalidTransferHookProgramId,
+            "expected {}, found {:?}",
+            program_id,
+            hook_program_id
+        );
+    } else {
+        return err!(StakeError::MissingTransferHook);
+    }
+
+    // 2. vault (token account)
+    // - must be initialized
+    // - have the vault signer (PDA) as owner
+    // - have the correct mint
+    // - amount equal to 0
+
+    let vault_data = ctx.accounts.vault.try_borrow_data()?;
+    // unpack checks if the token is initialized
+    let vault = PodStateWithExtensions::<PodAccount>::unpack(&vault_data)?;
+
+    let (vault_signer, bump) = Pubkey::find_program_address(
+        &["token-owner".as_bytes(), ctx.accounts.config.key.as_ref()],
+        program_id,
+    );
+
+    require!(
+        vault.base.owner == vault_signer,
+        StakeError::InvalidTokenOwner,
+        "vault"
+    );
+
+    require!(
+        &vault.base.mint == ctx.accounts.mint.key,
+        StakeError::InvalidMint,
+        "vault"
+    );
+
+    let amount: u64 = vault.base.amount.into();
+    require!(amount == 0, StakeError::AmountGreaterThanZero, "vault");
+
+    // 3. config
+    // - owner must be stake program
+    // - have the correct length
+    // - be uninitialized
+
+    let mut data = ctx.accounts.config.try_borrow_mut_data()?;
+
+    require!(
+        ctx.accounts.config.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "config"
+    );
+
+    require!(
+        data.len() == Config::LEN,
+        StakeError::InvalidAccountDataLength,
+        "config"
+    );
+
+    let config = bytemuck::from_bytes_mut::<Config>(&mut data);
+
+    require!(
+        !config.is_initialized(),
+        ProgramError::AccountAlreadyInitialized,
+        "config"
+    );
```

Comment:
This is currently fine, but can be dangerous -- if I pass in a different initialized account type with the same length as a config account, the account's discriminator will be set to something else, and pass this check.

Typically, you want to make sure that the discriminator is uninitialized (ie all 0s) to be sure that *nothing* has been initialized here.

Reply:
Good point! I added an `is_uninitialized()` helper to the `Config` account that compares against an empty discriminator and used it here.